clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -424,9 +424,59 @@ caseCase _ e = return e
 {-# SCC caseCase #-}
 
 -- | Inline function with a non-representable result if it's the subject
--- of a Case-decomposition
+-- of a Case-decomposition. It's a custom topdown traversal that -for efficiency
+-- reasons- does not explore alternative of cases whose subject triggered an
+-- 'inlineNonRepWorker'.
 inlineNonRep :: HasCallStack => NormRewrite
-inlineNonRep _ e@(Case scrut altsTy alts)
+inlineNonRep _ctx = go
```

Comment:
I think this could be simplified to:

```haskell
inlineNonRepTraverse ctx e = do
  (e1, hasChanged) <- listen (inlineNonRep ctx e)
  case e1 of
    Case {} -> return e1
    _ -> allR inlineNonRepTraverse ctx e1
```

where `inlineNonRep` is what you call `inlineNonRepWorker`. Or am I missing something?

Reply:
I don't think it's equivalent, did you mean:

```haskell
inlineNonRepTraverse ctx e = do
  r <- listen (inlineNonRep ctx e)
  case r of
    (e1, Monoid.getAny -> True) -> return e1
    (e1, _) -> allR inlineNonRepTraverse ctx e1
```

? Or more directly: did you mean to ignore `hasChanged`? Why would you match on `Case`?

Reply:
Yes, sorry, I meant:
```haskell
inlineNonRepTraverse ctx e = do
  (e1, hasChanged) <- listen (inlineNonRep ctx e)
  case Monoid.getAny hasChanged of
    True -> return e1
    _ -> allR inlineNonRepTraverse ctx e1
```

Reply:
Yes, that's sensible. As discussed in a video call, this has the slight downside of keeping track of the context - even though it isn't used by the worker function.

Reply:
We would still need to special case the `Case`. I've changed the transformation to use `allR`, but retain the original behavior.